### PR TITLE
`Development`: Remove function calls in diff report template

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The template of the diff report viewer was written when avoiding function calls in the template wasn't a priority. This is against our client coding guidelines and can decrease the performance of our app.

### Description

The corresponding information is now pre-computed in ngOnInit, and I have removed the function calls from the template.

### Steps for Testing

Prerequisites:
- LocalVC test server
- 1 Instructor
- 1 Programming exercise with a participation

1. Log in to Artemis
2. Navigate to your programming exercise > Instructor actions: Participations > Open repository (of the participation)
3. Open the commit history and click on a commit
4. Ensure the commit is displayed correctly and that the files are listed in the correct (alphabetical) order.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

**Review by Jan-Thurner — DISMISSED** [review]
Tested on TS1. Works as expected.

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts [hunk 1]
```diff
@@ -108,12 +116,16 @@ export class GitDiffReportComponent implements OnInit {
...
-            }
-        });
-        this.nothingToDisplay = Object.keys(this.diffsReadyByPath).length === 0;
+        this.diffInformationForPaths = this.filePaths
+            .filter((path) => this.entriesByPath.get(path)?.length)
+            .map((path) => {
+                // entries is not undefined due to the filter
+                const entries = this.entriesByPath.get(path)!;
+                const templateFileContent = this.templateFileContentByPath.get(this.renamedFilePaths[path] || path);
+                const solutionFileContent = this.solutionFileContentByPath.get(path) ?? '';
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts:125]: Is there a reason why you use `||` in one occasion but `??` in the other?

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me. Left one question:

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts [hunk 2]
```diff
@@ -123,7 +135,14 @@ export class GitDiffReportComponent implements OnInit {
...
      */
     onDiffReady(path: string, ready: boolean) {
-        this.diffsReadyByPath[path] = ready;
-        this.allDiffsReady = Object.values(this.diffsReadyByPath).reduce((a, b) => a && b);
+        const index = this.diffInformationForPaths.findIndex((info) => info.path === path);
+        if (index !== -1) {
+            this.diffInformationForPaths[index].diffReady = ready;
+            this.allDiffsReady = Object.values(this.diffInformationForPaths)
+                .map((info) => info.diffReady)
+                .reduce((a, b) => a && b);
```
Comment (SimonEntholzer) [inline src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts:143]:
```suggestion
this.allDiffsReady = Object.values(this.diffInformationForPaths)
    .every((info) => info.diffReady);
```
using this e.g. could make the code more concise here

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Code looks good, added a small suggestion

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts (hunk 1 above)
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts]: The `??` was actually not even required here (`undefined` is a valid value for the solutionFileContent). I also changed the `renamedFilePaths` s.t. typescript now acknowledges that `renamedFilePaths[path]` can be undefined and replaced `||` with `??`.

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts (hunk 2 above)
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts]: Added 👍

File: src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts [hunk 3]
```diff
@@ -108,12 +116,16 @@ export class GitDiffReportComponent implements OnInit {
...
-            if (this.entriesByPath.get(path)?.length) {
-                this.diffsReadyByPath[path] = false;
-            }
-        });
-        this.nothingToDisplay = Object.keys(this.diffsReadyByPath).length === 0;
+        this.diffInformationForPaths = this.filePaths
+            .filter((path) => this.entriesByPath.get(path)?.length)
+            .map((path) => {
+                // entries is not undefined due to the filter
+                const entries = this.entriesByPath.get(path)!;
```
[the remark this answers is not in the document]
Comment (pzdr7) [reply src/main/webapp/app/exercises/programming/hestia/git-diff-report/git-diff-report.component.ts:123]: Not the case here

**Review by BBesrour — COMMENTED** [review]
Tested on TS1, works as expected

**Review by BBesrour — APPROVED** [review]

**Review by SimonEntholzer — APPROVED** [review]
code lgtm now

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
Tested on TS1. Works as expected.

The server Codestyle check currently fails.

**Comment by pzdr7** [comment]
> Tested on TS1. Works as expected.
> 
> The server Codestyle check currently fails.

There are no server changes in this pull request

**Review by Jan-Thurner — APPROVED** [review]
As this PR does not contain any server changes and it works as expected, I approve this PR.

**Review by JohannesStoehr — APPROVED** [review]

**Review by egekurt123 — APPROVED** [review]
Code LGTM and works as expected on TS1

**Review by rabeatwork — APPROVED** [review]
Maintainer approved

**Review by sarpsahinalp — APPROVED** [review]
code lgtm

**Pull request merged**